src/Neo/Cryptography/Crypto.cs

```diff
@@ -112,6 +108,107 @@ public static byte[] Sign(byte[] message, byte[] priKey, ECC.ECCurve ecCurve = n
             return ecdsa.SignData(message, hashAlg);
         }
 
+        /// <summary>
+        /// ECRecover
+        /// </summary>
+        /// <param name="curve">Curve</param>
+        /// <param name="signature">Signature</param>
+        /// <param name="hash">Message hash</param>
+        /// <param name="format">Signature format</param>
+        /// <returns>Allowed Public keys</returns>
+        public static ECC.ECPoint[] ECRecover(ECC.ECCurve curve, byte[] signature, byte[] hash, SignatureFormat format = SignatureFormat.Der)
+        {
+            BigInteger r, s;
+            int recId = 0, recIdTo = 4;
+
+            // Decode signature
+
+            switch (format)
+            {
+                case SignatureFormat.Der:
+                    {
+                        var derSequence = (DerSequence)Asn1Object.FromByteArray(signature);
+                        r = ((DerInteger)derSequence[0]).Value;
+                        s = ((DerInteger)derSequence[1]).Value;
+
+                        if (derSequence.Count == 3)
+                        {
+                            recId = ((DerInteger)derSequence[2]).IntValueExact;
+                            recIdTo = recId + 1;
+                        }
+                        break;
+                    }
+                case SignatureFormat.Fixed32:
+                    {
+                        r = new(1, signature, 0, 32);
+                        s = new(1, signature, 32, 32);
+
+                        if (signature.Length == RecuperableSignatureLength)
+                        {
+                            recId = signature[SignatureLength];
+                            recIdTo = recId + 1;
+                        }
+                        break;
+                    }
+                default: throw new InvalidOperationException("Invalid signature format");
+            }
+
+            // Validate values
```

Comment:
on the topic of validation, maybe having the s value in the lower half of the domain would be a good guard against signature maleability. At least that's how that was addressed in ethereum where the wallets only generate (r,s) pairs with s lower than n/2 (where n is the order of the curve). Because a pair (r, n-s) is also a valid solution leading to possible issues.